keps/sig-storage/0033-20181111-transfer-pvc-between-namespaces.md

```diff
@@ -0,0 +1,179 @@
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [Proposal](#proposal)
+    * [User Stories [optional]](#user-stories-optional)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+    * [Risks and Mitigations](#risks-and-mitigations)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+* [Drawbacks [optional]](#drawbacks-optional)
+* [Alternatives [optional]](#alternatives-optional)
+
+[Tools for generating]: https://github.com/ekalinin/github-markdown-toc
+
+## Summary
+
+Provide a mechanism to allow users to "give" a PVC to a user in another namespace based on mutual acceptance. A transfer should be initiated by updating an existing PVC with a transfer request annotation, and accepted by the intended destination namespace by the recipient requesting a PVC using the standard mechanism and specifying the correct PVC Name and transfer annotation as per the transfer request.
+
+* Provide a Cluster wide mechanism to enable or disable the transfer of PVCs
+* Expose standardized process for transferring a PVC from one Namespace to another
+* Require explicit acceptance of a PVC into another users Namespace, ensuring a malitious user can't just dump PVCs into another users Namespace
+* Utilize existing PVC creation semantics for creating a PVC from a Transfer
+* Honor existing configuration options including quota checks, Storage Class access etc
+
+The process of transferring a PVC/Volume is as follows:
+1. Original user indicates they're willing to ``give`` the volume to another namespace
+2. The receiving user indicates they'd like to ``accept`` the volume into their namespace
+3. When the original user deletes the PVC from their namespace, the claim request from the receiving user is fulfilled
+
+## Motivation
+
+There are a number of uses cases where a user would like to have the ability to transfer an existing PVC from one namespace to another.  This is a valuable workflow for persistent storage and enables the ability to easily duplicate and transfer data sets from one environment to another.  These populated PVCs could be a clone of another volume, a volume from snapshot, or data that was written to the volume via an application (ie a database).
+
+An example use case for this feature would be a cluster segmented into two namespaces; namespace-a for production staging, and namespace-b for production.  There are cases where an application could be developed and tested with the same production data without risking any modification or corruption of data in both environments.  Rather than reproducing the data in both namespaces, it would be much more efficient to be able to clone or restore the data from a snapshot in to a volume and then transfer that new volume to the desired namespace.
```

Comment:
While I appreciate this use case, I'm not sure this is a good behavior. For example, in this setup the staging environment is no longer available so the staging wasn't really staging it was pre-launch prod. Why wouldn't this just be the production namespace in the first place?

Wouldn't a better approach be to copy the staging data, as a one time task, from stating to production? Then, wouldn't you want periodic tasks that copy the production data back to staging for continued dev leveraging staging? This copying is far different from transferring.

I'm just thinking out loud but this example does not sound like a good case to justify the behavior. Although, I might be missing something and am happy hear about it.

Reply:
Yeah, 

> Wouldn't a better approach be to copy the staging data, as a one time task, from stating to production? Then, wouldn't you want periodic tasks that copy the production data back to staging for continued dev leveraging staging? This copying is far different from transferring.

Sure, but that's a different workflow IMO (and I'd still prefer to clone and transfer the new PVC in that case anyway), maybe the wording isn't great on my part; the point here was that data isn't static in either case.  Say for example you have a production env that is encountering errors, you try and test things in your test env/namespace but it turns out that the issue is dependent upon data (which isn't uncommon); the scenario here allows a method whereby you can reproduce data if needed.

I'd also argue that it's a more efficient and secure method of moving from staging to production as well; by providing a mechanism to easily and completely duplicate the staging env in another namespace (prod or otherwise) without manually moving data which would require poor security settings (allowing both namespaces access to the raw data) which can be extremely time consuming.  Given most storage backends will have the ability to quickly and effeciently clone a volume, this is a much easier way to replicate and transfer that data between namespaces.

Reply:
Thanks for explaining the workflow you are talking about.

If this is your workflow, would it be better to copy a snapshot from one namespace to another so it can be restored there? That way no PV/PVC is within the production namespace.

Reply:
> Thanks for explaining the workflow you are talking about.
> 
> If this is your workflow, would it be better to copy a snapshot from one namespace to another so it can be restored there? That way no PV/PVC is within the production namespace.

The problem with transferring snapshots is that it opens up a considerable amount of complexity.  Many storage devices link their snapshots to their parent volumes, the result being that we would have cross namespace dependencies on resources which IMO is pretty ugly to manage and frankly not worthwhile.  Creating a volume from a snapshot and then transferring it means that everything in an end users namespace remains independent and under a single namespace control.